File: LeetCode/leet_117/connect_node.cpp
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};

class Solution {
public:
    Node* connect(Node* root) {
        Node* dummyHead = new Node();
        while (root != nullptr) {
            Node* cur = dummyHead;
            while (root != nullptr) {
                if (root->left) {
                    cur->next = root->left;
                    cur = cur->next;
                }

                if (root->right) {
                    cur->next = root->right;
                    cur = cur->next;
                }
                root = root->next;
            }
            root = dummyHead->next;
            dummyHead = nullptr;
        }
    }
};

File: LeetCode/leet_117/connect_node.java
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};

class Solution {
    public Node connect(Node root) {
       Node tempChild = new Node(0);
           while (root != null) {
               Node currentChild = tempChild;
               while (root != null) {
                   if (root.left != null) {
                       currentChild.next = root.left;
                       currentChild = currentChild.next;
                   }
                   if (root.right != null) {
                       currentChild.next = root.right;
                       currentChild = currentChild.next;
                   }
                   root = root.next;
               }
               root = tempChild.next;
               tempChild.next = null;
           }
    }
}
